### PR TITLE
Issue 17

### PR DESCRIPTION
**Pull request by awenger**

### CODE REVIEW

**Comment by nlf** [comment]
I would prefer that this return an actual error object instead of json with an errmsg property. The signature for the connect callback should be `function (err)` where err is undefined if no error occurred, and an error object if one did occur.

Additionally, I suspect a 500ms timeout may be too aggressive for some situations. Maybe a 1 second timeout default and make it configurable?

**Comment by awenger** [comment]
I just wanted to be consistent with the rest of the api. But `function(err)` is also ok for me.
True 500 ms could be a little bit agressive. And configurable is always good. I'll change it.

**Comment by awenger** [comment]
I also added a test for the timout functionality

File: index.js [hunk 1]
```diff
@@ -262,11 +264,21 @@ RiakPBC.prototype.ping = function (callback) {
...
 
 RiakPBC.prototype.connect = function (callback) {
-    if (this.connected) return callback();
+    if (this.connected) return callback(null);
+
     var self = this;
+
+    var timeoutGuard = setTimeout(function(){
+        timeoutGuard = null;
+        callback('timeout');
```
Comment (nlf) [inline index.js]: `new Error('Connection timeout')` is what I had in mind

**Comment by nlf** [comment]
Aside from the one nit pick, this looks great. Thanks for including a test too!

**Comment by awenger** [comment]
like this?

**Comment by awenger** [comment]
I found something else. I don't know whether it's a real bug. The Following code will cause two timeout Errors:

```
var api = require('riakpbc').createClient();

api.connect( function(reply){
  console.log(reply);
});

api.getBuckets(function(reply){
  console.log(reply);
});
```

It also happens with the current version of riakpbc (there of course no Errors, only the cb is never called). So i think it's not caused by the recent changes. Should I open another bug ticket?

**Comment by nlf** [comment]
That's expected, when you run a command without running connect first - connect is called for you in the background. What I'll need to do is reorganize everything to follow the `function (err, reply)` signature and pass error objects through.. This will need a bit of rethinking..

**Comment by awenger** [comment]
For my own project I wrote a small wrapper. But I also think it would be nice to have this cb style in riakpbc

```
var riakpbc = require('riakpbc');

function formatReply(reply, cb){
  if(reply.errmsg)
    return cb(reply.errmsg);
  return cb(null, reply);
}


module.exports = function create(conf){
  var client = riakpbc.createClient(conf);

  return {
    getBuckets: function(cb){
      client.getBuckets(function(reply){
        formatReply(reply, cb);
      });
    },
    getBucket: function(params, cb){
      client.getBucket(params, function(reply){
        formatReply(reply, cb);
      });
    },
    setBucket: function(params, cb){
      client.setBucket(params, function(reply){
        formatReply(reply, cb);
      });
    },
    resetBucket: function(params, cb){
      client.resetBucket(params, function(reply){
        formatReply(reply, cb);
      });
    },
    getKeys: function(params, cb){
      client.getKeys(params, function(reply){
        formatReply(reply, cb);
      });
    },
    streamKeys: function(params){
      return client.getKeys(params, true);
    },
    get: function(params, cb){
      client.get(params, function(reply){
        formatReply(reply, cb);
      });
    },
    put: function(params, cb){
      client.put(params, function(reply){
        formatReply(reply, cb);
      });
    },
    del: function(params, cb){
      client.del(params, function(reply){
        formatReply(reply, cb);
      });
    },
    updateCounter: function(params, cb){
      client.updateCounter(params, function(reply){
        formatReply(reply, cb);
      });
    },
    getCounter: function(params, cb){
      client.getCounter(params, function(reply){
        formatReply(reply, cb);
      });
    },
    mapred: function(params, cb){
      client.mapred(params, function(reply){
        formatReply(reply, cb);
      });
    },
    getIndex: function(query, cb){
      client.getIndex(query, function(reply){
        formatReply(reply, cb);
      });
    },
    search: function(params, cb){
      client.search(params, function(reply){
        formatReply(reply, cb);
      });
    },
    ping: function(cb){
      client.ping(function(reply){
        formatReply(reply, cb);
      });
    },
    setClientId: function(params, cb){
      client.setClientId(params, function(reply){
        formatReply(reply, cb);
      });
    },
    getClientId: function(cb){
      client.getClientId(function(reply){
        formatReply(reply, cb);
      });
    },
    getServerInfo: function(cb){
      client.getServerInfo(function(reply){
        formatReply(reply, cb);
      });
    },
    connect: function(cb){
      client.connect(function(err){
        cb(err);
      });
    },
    disconnect: function(){
      client.disconnect();
    }
  };
};
```

**Comment by nlf** [comment]
I'm working on patching the main client to parse replies with an errmsg property as error objects and do proper err first callbacks

**Comment by nlf** [comment]
Merging this as a start, and opened #19 to track the main issue. Once that ticket is closed this will get published with it.

**Pull request merged**